### PR TITLE
Implement support for forward declarations of variables

### PR DESCRIPTION
**Pull request by simon-camp**

### CODE REVIEW

File: include/emitc/Target/Cpp.h [hunk 1]
```diff
@@ -86,6 +86,9 @@ struct CppEmitter {
   /// the array size.
   LogicalResult emitTupleType(ArrayRef<Type> types);
 
+  /// Emits an assignment for a variable which has been declared previously.
+  LogicalResult emitVariableAssignment(OpResult result);
+
   /// Emits a variable declaration for a result of an operation.
   LogicalResult emitVariableDeclaration(OpResult result,
                                         bool trailingSemicolon);
```
Comment (marbre) [inline include/emitc/Target/Cpp.h:94]: Do we want to reintroduce the default `= true` here, if it is removed from `emitOperation`?

File: lib/Target/Cpp/TranslateToCpp.cpp [hunk 2]
```diff
@@ -28,26 +28,34 @@ using llvm::formatv;
 static LogicalResult printConstantOp(CppEmitter &emitter,
                                      ConstantOp constantOp) {
   auto &os = emitter.ostream();
-  if (failed(emitter.emitVariableDeclaration(
-          constantOp.getOperation()->getResult(0),
-          /*trailingSemicolon=*/false)))
-    return failure();
+  auto result = constantOp.getOperation()->getResult(0);
+  auto value = constantOp.value();
```
Comment (marbre) [inline lib/Target/Cpp/TranslateToCpp.cpp:32]: Hm, any chance to not loose the check?

File: lib/Target/Cpp/TranslateToCpp.cpp [hunk 3]
```diff
@@ -28,26 +28,34 @@ using llvm::formatv;
...
+  // Emit an assignment if variables are forward declared.
+  if (emitter.forwardDeclaredVariables()) {
+    if (failed(emitter.emitVariableAssignment(result)))
+      return failure();
+  } else {
+    if (failed(emitter.emitVariableDeclaration(result,
+                                               /*trailingSemicolon=*/false)))
+      return failure();
+    if (emitter.restrictedToC())
+      os << " = ";
```
Comment (marbre) [inline lib/Target/Cpp/TranslateToCpp.cpp:52]: Why not
```suggestion
    if (emitter.restrictedToC()) {
      os << " = ";
    } else {
      os << "{";
    }
```
Makes somehow more sense to me to put this here and to remove below. At least at this late hour ;)

**Review by marbre — CHANGES_REQUESTED** [review]
Some first suggestions. Still need to go through the if and for op printer.

File: include/emitc/Target/Cpp.h (hunk 1 above)
Comment (simon-camp) [reply include/emitc/Target/Cpp.h:94]: I'd vote for no. Having to think about this for each operation is a good thing.

**Comment by simon-camp** [comment]
I think this is feature complete for now. Tests for `emitc.call, emitc.const,emitc.getaddressof, !emitc.opaque`, need to be added/refactored, so they are run on all (applicable) target combinations.

File: include/emitc/Target/Cpp.h (hunk 1 above)
Comment (marbre) [reply include/emitc/Target/Cpp.h:94]: I am fine with this :+1:

**Review by marbre — APPROVED** [review]
Awesome work Simon!

**Pull request merged**